diff -ruN orig/arbtt.cabal new/arbtt.cabal
--- orig/arbtt.cabal	2014-08-10 08:32:58.666725004 +0300
+++ new/arbtt.cabal	2014-08-10 08:32:58.000000000 +0300
@@ -35,7 +35,7 @@
     build-depends:
         base == 4.5.* || == 4.6.* || == 4.7.*,
         filepath, directory, transformers, time >= 1.4, utf8-string,
-        aeson == 0.6.* || == 0.7.*,
+        aeson,
         array == 0.4.* || == 0.5.*,
         binary >= 0.5,
         bytestring, deepseq, strict, old-locale
@@ -82,7 +82,7 @@
         binary >= 0.5,
         deepseq, bytestring, utf8-string, time >= 1.4, strict,
         transformers, unix, directory, filepath,
-        aeson == 0.6.* || == 0.7.*,
+        aeson,
         array == 0.4.* || == 0.5.*,
         terminal-progress-bar, bytestring-progress
     other-modules:
@@ -112,7 +112,7 @@
         base == 4.5.* || == 4.6.* || == 4.7.*,
         parsec == 3.*,
         containers == 0.5.*,
-        aeson == 0.6.* || == 0.7.*,
+        aeson,
         array == 0.4.* || == 0.5.*,
         binary >= 0.5,
         deepseq, bytestring, utf8-string, time >= 1.4, strict,
