diff -ru orig/async.cabal new/async.cabal
--- orig/async.cabal	2013-12-09 14:04:55.984162531 +0200
+++ new/async.cabal	2013-12-09 14:04:55.000000000 +0200
@@ -70,13 +70,13 @@
 
 library
     exposed-modules:     Control.Concurrent.Async
-    build-depends:       base >= 4.3 && < 4.7, stm >= 2.2 && < 2.5
+    build-depends:       base >= 4.3 && < 4.8, stm >= 2.2 && < 2.5
 
 test-suite test-async
     type:       exitcode-stdio-1.0
     hs-source-dirs: test
     main-is:    test-async.hs
-    build-depends: base >= 4.3 && < 4.7,
+    build-depends: base >= 4.3 && < 4.8,
                    async,
                    test-framework,
                    test-framework-hunit,
