diff -ruN orig/aws.cabal new/aws.cabal
--- orig/aws.cabal	2014-08-27 18:53:03.893146895 +0300
+++ new/aws.cabal	2014-08-27 18:53:03.000000000 +0300
@@ -118,7 +118,7 @@
                        bytestring           >= 0.9     && < 0.11,
                        case-insensitive     >= 0.2     && < 1.3,
                        cereal               >= 0.3     && < 0.5,
-                       conduit              >= 1.1     && < 1.2,
+                       conduit              >= 1.1     && < 1.3,
                        conduit-extra        >= 1.1     && < 1.2,
                        containers           >= 0.4,
                        cryptohash           >= 0.11     && < 0.12,
