diff -ru orig/aws.cabal new/aws.cabal
--- orig/aws.cabal	2014-04-02 13:40:50.015820737 +0300
+++ new/aws.cabal	2014-04-02 13:40:49.000000000 +0300
@@ -130,7 +130,9 @@
                        unordered-containers >= 0.2,
                        utf8-string          == 0.3.*,
                        vector               >= 0.10,
-                       xml-conduit          >= 1.1 && <1.2
+                       xml-conduit          >= 1.1 && <1.2,
+                       http-client < 0.3,
+                       ghc-prim
 
   GHC-Options: -Wall
 
