diff -ruN orig/BlogLiterately.cabal new/BlogLiterately.cabal
--- orig/BlogLiterately.cabal	2014-08-20 18:59:49.878397566 +0300
+++ new/BlogLiterately.cabal	2014-08-20 18:59:49.000000000 +0300
@@ -64,7 +64,7 @@
                    pandoc-citeproc >= 0.1.2 && < 0.5,
                    highlighting-kate >= 0.5 && < 0.6,
                    data-default >= 0.5 && < 0.6,
-                   lens >= 3.8 && < 4.3
+                   lens >= 3.8
   Exposed-modules: Text.BlogLiterately
                    Text.BlogLiterately.Block
                    Text.BlogLiterately.Ghci
