diff -ruN orig/BlogLiterately.cabal new/BlogLiterately.cabal
--- orig/BlogLiterately.cabal	2014-08-10 08:43:22.242711731 +0300
+++ new/BlogLiterately.cabal	2014-08-10 08:43:22.000000000 +0300
@@ -63,7 +63,7 @@
                    pandoc-citeproc >= 0.1.2 && < 0.4,
                    highlighting-kate >= 0.5 && < 0.6,
                    data-default >= 0.5 && < 0.6,
-                   lens >= 3.8 && < 4.3
+                   lens
   Exposed-modules: Text.BlogLiterately
                    Text.BlogLiterately.Block
                    Text.BlogLiterately.Ghci
