diff -ru orig/bson.cabal new/bson.cabal
--- orig/bson.cabal	2014-03-13 08:24:42.121883180 +0200
+++ new/bson.cabal	2014-03-13 08:24:41.000000000 +0200
@@ -26,7 +26,7 @@
   Build-depends:      base < 5
                     , time
                     , bytestring
-                    , binary == 0.7.*
+                    , binary
                     , cryptohash
                     , data-binary-ieee754
                     , mtl >= 2
