diff -ru orig/bson.cabal new/bson.cabal
--- orig/bson.cabal	2014-02-19 18:55:36.860771700 +0200
+++ new/bson.cabal	2014-02-19 18:55:36.000000000 +0200
@@ -26,7 +26,7 @@
   Build-depends:      base < 5
                     , time
                     , bytestring
-                    , binary == 0.7.*
+                    , binary >= 0.5.1.0 && < 0.8
                     , cryptohash
                     , data-binary-ieee754
                     , mtl >= 2
@@ -51,7 +51,7 @@
                     , base < 5
                     , time
                     , bytestring
-                    , binary == 0.7.*
+                    , binary >= 0.5.1.0 && < 0.8
                     , cryptohash
                     , data-binary-ieee754
                     , mtl >= 2
