diff -ru orig/tests/doctests.hsc new/tests/doctests.hsc
--- orig/tests/doctests.hsc	2013-09-02 08:40:59.176527818 +0300
+++ new/tests/doctests.hsc	2013-09-02 08:40:59.000000000 +0300
@@ -59,7 +59,8 @@
   : "-optPdist/build/autogen/cabal_macros.h"
   : "-hide-all-packages"
   : "-Iincludes"
-  : map ("-package="++) deps ++ sources
+  : "dist/build/cbits/i2d.o"
+  : map ("-package="++) (filter (not . ("bytes-" `isPrefixOf`)) deps) ++ sources
 
 getSources :: IO [FilePath]
 getSources = filter (isSuffixOf ".hs") <$> go "src"
