diff -ru orig/concurrent-extra.cabal new/concurrent-extra.cabal
--- orig/concurrent-extra.cabal	2014-04-04 14:42:42.732063525 +0300
+++ new/concurrent-extra.cabal	2014-04-04 14:42:42.000000000 +0300
@@ -50,7 +50,7 @@
 -------------------------------------------------------------------------------
 
 library
-  build-depends: base                 >= 3       && < 4.7
+  build-depends: base                 >= 3       && < 4.8
                , base-unicode-symbols >= 0.1.1   && < 0.3
                , stm                  >= 2.1.2.1 && < 2.5
                , unbounded-delays     >= 0.1     && < 0.2
@@ -80,7 +80,7 @@
 
   ghc-options: -Wall -threaded
 
-  build-depends: base                 >= 3       && < 4.7
+  build-depends: base                 >= 3       && < 4.8
                , base-unicode-symbols >= 0.1.1   && < 0.3
                , stm                  >= 2.1.2.1 && < 2.5
                , unbounded-delays     >= 0.1     && < 0.2
