diff -ru orig/csv-conduit.cabal new/csv-conduit.cabal
--- orig/csv-conduit.cabal	2014-04-03 10:44:52.994206357 +0300
+++ new/csv-conduit.cabal	2014-04-03 10:44:52.000000000 +0300
@@ -78,6 +78,7 @@
     , base >= 4 && < 5
     , bytestring
     , conduit >= 1.0 && < 2.0
+    , conduit-extra
     , containers >= 0.3
     , monad-control
     , text
@@ -90,6 +91,7 @@
     , mtl
     , mmorph
     , primitive
+    , resourcet
   ghc-prof-options: -fprof-auto
 
   if impl(ghc >= 7.2.1)
diff -ru orig/src/Data/CSV/Conduit.hs new/src/Data/CSV/Conduit.hs
--- orig/src/Data/CSV/Conduit.hs	2014-04-03 10:44:52.962206360 +0300
+++ new/src/Data/CSV/Conduit.hs	2014-04-03 10:44:52.000000000 +0300
@@ -36,6 +36,9 @@
 import           Control.Monad.Primitive
 import           Control.Monad.ST
 import           Control.Monad.Trans
+import           Control.Monad.Trans.Resource       (MonadResource, MonadThrow,
+                                                     runExceptionT,
+                                                     runResourceT)
 import           Data.Attoparsec.Types              (Parser)
 import qualified Data.ByteString                    as B
 import           Data.ByteString.Char8              (ByteString)
