diff -ruN orig/diagrams-builder.cabal new/diagrams-builder.cabal
--- orig/diagrams-builder.cabal	2014-08-10 08:43:22.462711726 +0300
+++ new/diagrams-builder.cabal	2014-08-10 08:43:22.000000000 +0300
@@ -57,7 +57,7 @@
                        split >= 0.2 && < 0.3,
                        haskell-src-exts >= 1.14 && < 1.16,
                        cmdargs >= 0.6 && < 0.11,
-                       lens >= 3.9 && < 4.3,
+                       lens,
                        hashable >= 1.1 && < 1.3,
                        exceptions >= 0.3 && < 0.7
   hs-source-dirs:      src
@@ -100,7 +100,7 @@
                        diagrams-lib >= 0.6 && < 1.3,
                        diagrams-cairo >= 0.6 && < 1.3,
                        cmdargs >= 0.6 && < 0.11,
-                       lens >= 3.8 && < 4.3
+                       lens
 
 executable diagrams-builder-svg
   main-is:             diagrams-builder-svg.hs
@@ -141,4 +141,4 @@
                        diagrams-lib >= 0.6 && < 1.3,
                        diagrams-postscript >= 0.6 && < 1.2,
                        cmdargs >= 0.6 && < 0.11,
-                       lens >= 3.8 && < 4.3
+                       lens
