diff -ru orig/distributed-process.cabal new/distributed-process.cabal
--- orig/distributed-process.cabal	2014-03-25 06:24:53.476644211 +0200
+++ new/distributed-process.cabal	2014-03-25 06:24:53.000000000 +0200
@@ -90,7 +90,7 @@
   Main-Is:           TestCH.hs
   Build-Depends:     base >= 4.4 && < 5,
                      random >= 1.0 && < 1.1,
-                     ansi-terminal >= 0.5 && < 0.6,
+                     ansi-terminal >= 0.5,
                      distributed-process,
                      network-transport >= 0.3 && < 0.4,
                      network-transport-tcp >= 0.3 && < 0.4,
@@ -111,7 +111,7 @@
   Main-Is:           TestClosure.hs
   Build-Depends:     base >= 4.4 && < 5,
                      random >= 1.0 && < 1.1,
-                     ansi-terminal >= 0.5 && < 0.6,
+                     ansi-terminal >= 0.5,
                      distributed-static >= 0.2 && < 0.3,
                      distributed-process,
                      network-transport >= 0.3 && < 0.4,
@@ -131,7 +131,7 @@
   Main-Is:           TestStats.hs
   Build-Depends:     base >= 4.4 && < 5,
                      random >= 1.0 && < 1.1,
-                     ansi-terminal >= 0.5 && < 0.6,
+                     ansi-terminal >= 0.5,
                      containers >= 0.4 && < 0.6,
                      stm >= 2.3 && < 2.5,
                      distributed-process,
