diff -ru orig/distributed-process-simplelocalnet.cabal new/distributed-process-simplelocalnet.cabal
--- orig/distributed-process-simplelocalnet.cabal	2014-03-27 18:23:44.960359467 +0200
+++ new/distributed-process-simplelocalnet.cabal	2014-03-27 18:23:44.000000000 +0200
@@ -33,7 +33,7 @@
                      network >= 2.3 && < 2.5,
                      network-multicast >= 0.0 && < 0.1,
                      data-accessor >= 0.2 && < 0.3,
-                     binary >= 0.5 && < 0.7,
+                     binary >= 0.5,
                      containers >= 0.4 && < 0.6,
                      transformers >= 0.2 && < 0.4,
                      network-transport >= 0.3 && < 0.4,
@@ -55,7 +55,7 @@
                        network >= 2.3 && < 2.5,
                        network-multicast >= 0.0 && < 0.1,
                        data-accessor >= 0.2 && < 0.3,
-                       binary >= 0.5 && < 0.7,
+                       binary >= 0.5,
                        containers >= 0.4 && < 0.6,
                        transformers >= 0.2 && < 0.4,
                        network-transport >= 0.3 && < 0.4,
