diff -ru orig/esqueleto.cabal new/esqueleto.cabal
--- orig/esqueleto.cabal	2014-04-03 08:31:02.238377300 +0300
+++ new/esqueleto.cabal	2014-04-03 08:31:01.000000000 +0300
@@ -91,7 +91,7 @@
     , containers
     , HUnit
     , QuickCheck
-    , hspec               >= 1.9
+    , hspec               >= 1.8
     , persistent-sqlite   >= 1.2 && < 1.4
     , persistent-template >= 1.2 && < 1.4
     , monad-control
