diff -ru orig/fb.cabal new/fb.cabal
--- orig/fb.cabal	2014-06-30 12:41:53.267551936 +0300
+++ new/fb.cabal	2014-06-30 12:41:52.000000000 +0300
@@ -76,7 +76,7 @@
     , data-default
     , http-types
     , http-conduit         >= 2.0     && < 2.2
-    , attoparsec           >= 0.10.4  && < 0.12
+    , attoparsec           >= 0.10.4  && < 0.13
     , unordered-containers
     , aeson                >= 0.5     && < 0.8
     , base16-bytestring    >= 0.1
