diff -ru orig/fb-persistent.cabal new/fb-persistent.cabal
--- orig/fb-persistent.cabal	2013-12-26 14:17:59.531602454 +0200
+++ new/fb-persistent.cabal	2013-12-26 14:17:59.000000000 +0200
@@ -32,7 +32,7 @@
   build-depends:
       base       >= 4     && < 5
     , fb         >= 0.13
-    , persistent == 1.2.*
+    , persistent >= 1.2   && < 1.4
     , cereal     >= 0.3   && < 0.5
     , time       == 1.4.*
     , text
