diff -ruN orig/force-layout.cabal new/force-layout.cabal
--- orig/force-layout.cabal	2014-08-10 08:43:22.898711717 +0300
+++ new/force-layout.cabal	2014-08-10 08:43:22.000000000 +0300
@@ -23,7 +23,7 @@
   build-depends:       base >= 4.2 && < 4.8,
                        vector-space >=0.7 && <0.9,
                        vector-space-points >= 0.1.1 && < 0.3,
-                       lens >= 3 && < 4.3,
+                       lens,
                        containers >=0.4 && < 0.6,
                        data-default-class >= 0.0.1 && < 0.1
   hs-source-dirs:      src
