diff -ru orig/Git/CmdLine.hs new/Git/CmdLine.hs
--- orig/Git/CmdLine.hs	2014-04-04 10:18:25.564401057 +0300
+++ new/Git/CmdLine.hs	2014-04-04 10:18:25.000000000 +0300
@@ -24,7 +24,7 @@
 import           Control.Monad.Trans.Class
 import           Control.Monad.Trans.Reader (ReaderT, runReaderT)
 import qualified Data.ByteString as B
-import           Data.Conduit hiding (MonadBaseControl)
+import           Data.Conduit
 import qualified Data.Conduit.List as CL
 import           Data.Foldable (for_)
 import           Data.Function
