diff -ru orig/gitlib-test.cabal new/gitlib-test.cabal
--- orig/gitlib-test.cabal	2014-04-04 06:49:19.204668116 +0300
+++ new/gitlib-test.cabal	2014-04-04 06:49:19.000000000 +0300
@@ -28,6 +28,7 @@
         , bytestring
         , failure              >= 0.2.0
         , conduit
+        , conduit-extra
         , monad-control        >= 0.3.1
         , tagged               >= 0.4.4
         , text                 >= 0.11.2
