diff -ru orig/Git/Repository.hs new/Git/Repository.hs
--- orig/Git/Repository.hs	2014-04-03 09:46:22.102281090 +0300
+++ new/Git/Repository.hs	2014-04-03 09:46:21.000000000 +0300
@@ -3,6 +3,7 @@
 import Control.Exception.Lifted
 import Control.Monad
 import Control.Monad.IO.Class
+import Control.Monad.Trans.Control (MonadBaseControl)
 import Data.Conduit
 import Git.Types
 import System.Directory
diff -ru orig/gitlib.cabal new/gitlib.cabal
--- orig/gitlib.cabal	2014-04-03 09:46:22.102281090 +0300
+++ new/gitlib.cabal	2014-04-03 09:46:21.000000000 +0300
@@ -43,6 +43,7 @@
         , base16-bytestring    >= 0.1.1.5
         , bytestring           >= 0.9.2.1
         , conduit              >= 1.0.0
+        , conduit-extra        >= 1.0.0
         , containers           >= 0.4.2.1
         , directory            >= 1.1.0.2
         , failure              >= 0.2.0.1
