diff -ru orig/haskell-names.cabal new/haskell-names.cabal
--- orig/haskell-names.cabal	2014-03-06 15:00:30.424709530 +0200
+++ new/haskell-names.cabal	2014-03-06 15:00:30.000000000 +0200
@@ -271,7 +271,7 @@
     , data-lens-template
     , tagged
     , traverse-with-class
-    , type-eq == 0.4
+    , type-eq >= 0.4
     , Cabal >= 1.14 && < 1.20
   Hs-source-dirs:       src
   Ghc-options:          -Wall -fno-warn-name-shadowing
