diff -ru orig/CHANGES.txt new/CHANGES.txt
--- orig/CHANGES.txt	2014-04-03 10:49:26.498200538 +0300
+++ new/CHANGES.txt	2014-04-03 10:49:26.000000000 +0300
@@ -1,5 +1,6 @@
 Changelog for Hoogle
 
+    #57, support QuickCheck-2.7
 4.2.29
     #55, if reading as UTF8 fails, explicitly try Latin1
     Add QuickCheck as a dependency
diff -ru orig/hoogle.cabal new/hoogle.cabal
--- orig/hoogle.cabal	2014-04-03 10:49:26.518200537 +0300
+++ new/hoogle.cabal	2014-04-03 10:49:26.000000000 +0300
@@ -46,6 +46,7 @@
         binary,
         bytestring >= 0.9,
         conduit >= 0.2,
+        resourcet,
         parsec >= 2.1,
         deepseq >= 1.1,
         text >= 0.11,
@@ -145,6 +146,7 @@
         transformers >= 0.2,
         uniplate >= 1.6,
         conduit >= 0.2,
+        resourcet,
         parsec >= 2.1,
         wai >= 1.1,
         warp >= 1.1,
diff -ru orig/src/General/Web.hs new/src/General/Web.hs
--- orig/src/General/Web.hs	2014-04-03 10:49:26.510200537 +0300
+++ new/src/General/Web.hs	2014-04-03 10:49:26.000000000 +0300
@@ -30,7 +30,8 @@
 
 import Blaze.ByteString.Builder(toLazyByteString)
 import Data.Conduit.List(consume)
-import Data.Conduit(($$),Flush,runResourceT,Flush(Chunk))
+import Data.Conduit(($$),Flush,Flush(Chunk))
+import Control.Monad.Trans.Resource (runResourceT)
 
 type Args = [(String, String)]
 
diff -ru orig/src/Test/BWT_FM.hs new/src/Test/BWT_FM.hs
--- orig/src/Test/BWT_FM.hs	2014-04-03 10:49:26.502200538 +0300
+++ new/src/Test/BWT_FM.hs	2014-04-03 10:49:26.000000000 +0300
@@ -3,11 +3,10 @@
 module Test.BWT_FM(bwt_fm) where
 
 import Test.General
-import Test.QuickCheck
 import General.BurrowsWheeler
 
 
 bwt_fm = do
     compress "tomorrow and tomorrow and tomorrow" === (31,"wwwdd  nnoooaatttmmmrrrrrrooo  ooo")
     decompress (31,"wwwdd  nnoooaatttmmmrrrrrrooo  ooo") === "tomorrow and tomorrow and tomorrow"
-    quickCheck $ \x -> decompress (compress x) == x
+    randCheck $ \x -> decompress (compress x) == x
diff -ru orig/src/Test/General.hs new/src/Test/General.hs
--- orig/src/Test/General.hs	2014-04-03 10:49:26.502200538 +0300
+++ new/src/Test/General.hs	2014-04-03 10:49:26.000000000 +0300
@@ -3,7 +3,7 @@
 
 import Control.Monad
 import qualified Data.ByteString as BS
-import Test.QuickCheck
+import Test.QuickCheck(Arbitrary(..), quickCheckWithResult, stdArgs, Testable, Result(..))
 
 
 instance Arbitrary BS.ByteString where
