diff -ru orig/HTF.cabal new/HTF.cabal
--- orig/HTF.cabal	2013-09-01 15:37:06.661829510 +0300
+++ new/HTF.cabal	2013-09-01 15:37:06.000000000 +0300
@@ -153,7 +153,7 @@
                     QuickCheck >= 2.3,
                     base == 4.*,
                     random >= 1.0,
-                    containers >= 0.5,
+                    containers >= 0.4.2.1,
                     process >= 1.0,
                     directory >= 1.0,
                     mtl >= 1.1 && < 2.2,
diff -ru orig/Test/Framework/XmlOutput.hs new/Test/Framework/XmlOutput.hs
--- orig/Test/Framework/XmlOutput.hs	2013-09-01 15:37:06.649829507 +0300
+++ new/Test/Framework/XmlOutput.hs	2013-09-01 15:37:06.000000000 +0300
@@ -1,3 +1,4 @@
+{-# LANGUAGE CPP #-}
 {- |
 
 See <http://pzolee.blogs.balabit.com/2012/11/jenkins-vs-junit-xml-format/>
@@ -16,7 +17,11 @@
 
 import qualified Data.ByteString.Lazy as BSL
 import qualified Data.List as List
+#if MIN_VERSION_containers(0, 5, 0)
 import qualified Data.Map.Strict as Map
+#else
+import qualified Data.Map as Map
+#endif
 import qualified Data.Text as T
 import Text.Printf
 
