diff -ruN orig/kan-extensions.cabal new/kan-extensions.cabal
--- orig/kan-extensions.cabal	2014-08-25 17:24:28.206498700 +0300
+++ new/kan-extensions.cabal	2014-08-25 17:24:28.000000000 +0300
@@ -51,6 +51,7 @@
     mtl           >= 2.0.1   && < 2.3,
     pointed       >= 4       && < 5,
     semigroupoids >= 4       && < 5,
+    tagged,
     transformers  >= 0.2     && < 0.5
 
   exposed-modules:
