Only in new: dist
diff -ru orig/language-javascript.cabal new/language-javascript.cabal
--- orig/language-javascript.cabal	2013-12-09 14:11:28.596175378 +0200
+++ new/language-javascript.cabal	2013-12-09 14:11:28.000000000 +0200
@@ -30,7 +30,7 @@
 
 Library
   Build-depends:     base             >= 4       && < 5
-                   , array            >= 0.3     && < 0.5
+                   , array            >= 0.3     && < 0.6
                    , mtl              >= 1.1     && < 2.9
                    , containers       >= 0.2     && < 0.6
                    , utf8-light       >= 0.4     && < 1.0
diff -ru orig/src/Language/JavaScript/Parser/Lexer.hs new/src/Language/JavaScript/Parser/Lexer.hs
--- orig/src/Language/JavaScript/Parser/Lexer.hs	2013-12-09 14:11:28.592175378 +0200
+++ new/src/Language/JavaScript/Parser/Lexer.hs	2013-12-09 14:11:27.000000000 +0200
@@ -1,4 +1,4 @@
-{-# LANGUAGE CPP,MagicHash #-}
+{-# LANGUAGE BangPatterns, CPP,MagicHash #-}
 {-# LINE 1 "src-dev/Language/JavaScript/Parser/Lexer.x" #-}
 
 
@@ -708,11 +708,15 @@
 
 	let
 		(base) = alexIndexInt32OffAddr alex_base s
-		((I# (ord_c))) = fromIntegral c
+		!((I# (ord_c))) = fromIntegral c
 		(offset) = (base +# ord_c)
 		(check)  = alexIndexInt16OffAddr alex_check offset
 		
+#if MIN_VERSION_base(4, 7, 0)
+		(new_s) = if (I# (offset >=# 0#) /= 0) && (I# (check ==# ord_c) /= 0)
+#else
 		(new_s) = if (offset >=# 0#) && (check ==# ord_c)
+#endif
 			  then alexIndexInt16OffAddr alex_table offset
 			  else alexIndexInt16OffAddr alex_deflt s
 	in
