diff -ru orig/lens.cabal new/lens.cabal
--- orig/lens.cabal	2014-05-16 09:14:13.877832405 +0300
+++ new/lens.cabal	2014-05-16 09:14:13.000000000 +0300
@@ -201,7 +201,7 @@
     primitive                 >= 0.4.0.1  && < 0.6,
     profunctors               >= 4        && < 5,
     reflection                >= 1.1.6    && < 2,
-    scientific                >= 0.2      && < 0.3,
+    scientific                >= 0.2      && < 0.4,
     semigroupoids             >= 4        && < 5,
     semigroups                >= 0.8.4    && < 1,
     split                     >= 0.2      && < 0.3,
