diff -ru orig/MFlow.cabal new/MFlow.cabal
--- orig/MFlow.cabal	2014-04-03 19:38:07.181525338 +0300
+++ new/MFlow.cabal	2014-04-03 19:38:07.000000000 +0300
@@ -104,7 +104,7 @@
                    utf8-string -any, wai >=2.0.0, case-insensitive -any,
                    http-types -any, conduit -any, text -any, parsec -any, warp -any,
                    warp-tls -any, random -any, blaze-html -any, blaze-markup -any,
-                   monadloc -any, clientsession ==0.9.0.3
+                   monadloc -any, clientsession ==0.9.0.3, conduit-extra
     exposed-modules: MFlow MFlow.Wai.Blaze.Html.All MFlow.Forms
                      MFlow.Forms.Admin MFlow.Cookies MFlow.Wai
                      MFlow.Forms.Blaze.Html MFlow.Forms.Test
