diff -ru orig/network-transport-tests.cabal new/network-transport-tests.cabal
--- orig/network-transport-tests.cabal	2014-03-25 06:24:53.648644213 +0200
+++ new/network-transport-tests.cabal	2014-03-25 06:24:53.000000000 +0200
@@ -24,7 +24,7 @@
                        bytestring >= 0.9 && < 0.11,
                        random >= 1.0 && < 1.1,
                        mtl >= 2.1 && < 2.2,
-                       ansi-terminal >= 0.5 && < 0.6
+                       ansi-terminal >= 0.5
   hs-source-dirs:      src
   ghc-options:         -Wall -fno-warn-unused-do-bind
   extensions:          CPP,
