diff -ru orig/Octree.cabal new/Octree.cabal
--- orig/Octree.cabal	2014-05-01 18:10:50.650819156 +0300
+++ new/Octree.cabal	2014-05-01 18:10:50.000000000 +0300
@@ -33,13 +33,13 @@
 
 Test-suite test_Octree
   Type:              exitcode-stdio-1.0
-  Build-depends:     base>=4.0 && < 4.7, AC-Vector >= 2.3.0, QuickCheck >= 2.4.0
+  Build-depends:     base>=4.0 && < 4.8, AC-Vector >= 2.3.0, QuickCheck >= 2.4.0
   Main-is:           tests/test_Octree.hs
 
 Test-suite readme
   type:           exitcode-stdio-1.0
   -- We have a symlink: README.lhs -> README.md
   main-is:        README.lhs
-  Build-depends:  base>=4.0 && < 4.7, AC-Vector >= 2.3.0, QuickCheck >= 2.4.0, markdown-unlit
+  Build-depends:  base>=4.0 && < 4.8, AC-Vector >= 2.3.0, QuickCheck >= 2.4.0, markdown-unlit
   ghc-options:    -pgmL markdown-unlit
 
