Only in new: dist
diff -ru orig/process-conduit.cabal new/process-conduit.cabal
--- orig/process-conduit.cabal	2014-03-30 12:27:40.781431440 +0300
+++ new/process-conduit.cabal	2014-03-30 12:27:40.000000000 +0300
@@ -35,6 +35,7 @@
                      , process          >= 1.0
                      , conduit          == 1.0.*
                      , shakespeare-text >= 1.0
+                     , shakespeare
 
   ghc-options:         -Wall
 
