diff -ru orig/retry.cabal new/retry.cabal
--- orig/retry.cabal	2014-05-11 11:28:24.319992770 +0300
+++ new/retry.cabal	2014-05-11 11:28:24.000000000 +0300
@@ -30,7 +30,7 @@
   exposed-modules:     Control.Retry
   build-depends:       
     base            ==4.*, 
-    exceptions      >= 0.5 && < 0.6,
+    exceptions      >= 0.5 && < 0.7,
     transformers,
     data-default
   hs-source-dirs:      src
