diff -ru orig/scientific.cabal new/scientific.cabal
--- orig/scientific.cabal	2014-03-06 14:57:06.880706336 +0200
+++ new/scientific.cabal	2014-03-06 14:57:06.000000000 +0200
@@ -47,8 +47,8 @@
 
   build-depends: scientific
                , base             >= 4.3   && < 4.8
-               , tasty            >= 0.3.1 && < 0.8
-               , tasty-smallcheck >= 0.2   && < 0.3
+               , tasty            >= 0.3.1 && < 0.9
+               , tasty-smallcheck >= 0.2   && < 0.9
                , smallcheck       >= 1.0   && < 1.2
                , text             >= 0.8   && < 1.2
 
