diff -ru orig/snap.cabal new/snap.cabal
--- orig/snap.cabal	2014-03-19 10:17:25.886213731 +0200
+++ new/snap.cabal	2014-03-19 10:17:25.000000000 +0200
@@ -154,7 +154,7 @@
     containers                >= 0.3      && < 0.6,
     directory                 >= 1.0      && < 1.3,
     directory-tree            >= 0.11     && < 0.12,
-    dlist                     >= 0.5      && < 0.7,
+    dlist                     >= 0.5      && < 0.8,
     errors                    >= 1.4      && < 1.5,
     filepath                  >= 1.1      && < 1.4,
     -- Blacklist bad versions of hashable
