diff -ru orig/Database/SQLite/Simple.hs new/Database/SQLite/Simple.hs
--- orig/Database/SQLite/Simple.hs	2014-02-14 14:38:24.411759783 +0200
+++ new/Database/SQLite/Simple.hs	2014-02-14 14:38:24.000000000 +0200
@@ -343,7 +343,7 @@
        | otherwise -> errorColumnMismatch (ColumnOutOfBounds col)
     Errors []  -> throwIO $ ConversionFailed "" "" "unknown error"
     Errors [x] ->
-      throw x `catch` (\e -> errorColumnMismatch (e :: ColumnOutOfBounds))
+      throw x `Control.Exception.catch` (\e -> errorColumnMismatch (e :: ColumnOutOfBounds))
     Errors xs  -> throwIO $ ManyErrors xs
   where
     errorColumnMismatch :: ColumnOutOfBounds -> IO r
