diff -ru orig/stm-conduit.cabal new/stm-conduit.cabal
--- orig/stm-conduit.cabal	2014-04-03 08:22:14.310388538 +0300
+++ new/stm-conduit.cabal	2014-04-03 08:22:14.000000000 +0300
@@ -29,9 +29,10 @@
       , stm-chans      >= 2.0 && < 3.1
       , cereal         >= 0.4.0.1
       , cereal-conduit >= 0.7.2
-      , conduit        == 1.0.*
+      , conduit        >= 1.0 && < 1.2
+      , conduit-extra  >= 1.0 && < 1.2
       , directory      >= 1.1
-      , resourcet      >= 0.3 && < 0.5
+      , resourcet      >= 0.3 && < 1.2
       , async          >= 2.0.1
       , monad-control  >= 0.3.2
       , monad-loops    >= 0.4.2
