diff -ruN orig/Test/Run.hs new/Test/Run.hs
--- orig/Test/Run.hs	2014-08-19 09:57:25.537902164 +0300
+++ new/Test/Run.hs	2014-08-19 09:57:25.000000000 +0300
@@ -27,6 +27,7 @@
 import Language.Haskell.TH.Desugar.Expand
 import Language.Haskell.TH.Desugar.Sweeten
 import Language.Haskell.TH
+import Language.Haskell.TH.Syntax (qRunIO)
 
 import Control.Monad
 
@@ -121,7 +122,7 @@
                     case (resK, lhs) of
                       (DStarK, [DVarT _]) -> [| True |]
                       _                                     -> do
-                        runIO $ do
+                        qRunIO $ do
                           putStrLn "Failed bug8884 test:"
                           putStrLn $ show dinfo
                         [| False |] )
