diff -ru orig/src/Graphics/UI/Threepenny/Internal/Types.hs new/src/Graphics/UI/Threepenny/Internal/Types.hs
--- orig/src/Graphics/UI/Threepenny/Internal/Types.hs	2013-12-24 06:45:49.129496370 +0200
+++ new/src/Graphics/UI/Threepenny/Internal/Types.hs	2013-12-24 06:45:48.000000000 +0200
@@ -37,7 +37,13 @@
 newtype ElementId = ElementId BS.ByteString
     deriving (Data,Typeable,Show,Eq,Ord)
 
-instance NFData ElementId where rnf (ElementId x) = rnf x
+instance NFData ElementId where
+    rnf (ElementId x) =
+#if MIN_VERSION_bytestring(0, 10, 0)
+        rnf x
+#else
+        BS.length x `seq` ()
+#endif
 
 type EventId  = String
 type Handlers = Map EventId (E.Handler EventData)
