diff -ru orig/uuid.cabal new/uuid.cabal
--- orig/uuid.cabal	2014-06-10 13:09:03.962534162 +0300
+++ new/uuid.cabal	2014-06-10 13:09:03.000000000 +0300
@@ -67,7 +67,7 @@
                        uuid,
                        bytestring >= 0.9 && < 1.1,
                        HUnit >=1.2 && < 1.3,
-                       QuickCheck >=2.4 && < 2.7,
+                       QuickCheck >=2.4 && < 2.8,
                        random >= 1.0.1 && < 1.1,
                        test-framework == 0.8.*,
                        test-framework-hunit == 0.3.*,
