diff -ru orig/vault.cabal new/vault.cabal
--- orig/vault.cabal	2013-12-09 14:04:56.244162539 +0200
+++ new/vault.cabal	2013-12-09 14:04:56.000000000 +0200
@@ -47,7 +47,7 @@
 
 Library
     hs-source-dirs:     src
-    build-depends:      base >= 4.5 && < 4.7,
+    build-depends:      base >= 4.5 && < 4.8,
                         containers >= 0.4 && < 0.6,
                         unordered-containers >= 0.2.3.0 && < 0.3,
                         hashable >= 1.1.2.5 && < 1.3
