diff -ru orig/websockets.cabal new/websockets.cabal
--- orig/websockets.cabal	2014-04-03 08:32:53.818374925 +0300
+++ new/websockets.cabal	2014-04-03 08:32:53.000000000 +0300
@@ -69,7 +69,7 @@
     binary            >= 0.5    && < 0.8,
     blaze-builder     >= 0.3    && < 0.4,
     bytestring        >= 0.9    && < 0.11,
-    case-insensitive  >= 0.3    && < 1.2,
+    case-insensitive  >= 0.3    && < 1.3,
     containers        >= 0.3    && < 0.6,
     io-streams        >= 1.1    && < 1.2,
     mtl               >= 2.0    && < 2.2,
@@ -105,7 +105,7 @@
     binary            >= 0.5    && < 0.8,
     blaze-builder     >= 0.3    && < 0.4,
     bytestring        >= 0.9    && < 0.11,
-    case-insensitive  >= 0.3    && < 1.2,
+    case-insensitive  >= 0.3    && < 1.3,
     containers        >= 0.3    && < 0.6,
     io-streams        >= 1.1    && < 1.2,
     mtl               >= 2.0    && < 2.2,
