diff -ru orig/websockets.cabal new/websockets.cabal
--- orig/websockets.cabal	2013-09-12 10:30:46.697755480 +0300
+++ new/websockets.cabal	2013-09-12 10:30:46.000000000 +0300
@@ -73,7 +73,7 @@
     blaze-builder            >= 0.3    && < 0.4,
     blaze-builder-enumerator >= 0.2    && < 0.3,
     bytestring               >= 0.9    && < 0.11,
-    case-insensitive         >= 0.3    && < 1.1,
+    case-insensitive         >= 0.3    && < 1.2,
     containers               >= 0.3    && < 0.6,
     enumerator               >= 0.4.13 && < 0.5,
     mtl                      >= 2.0    && < 2.2,
@@ -106,7 +106,7 @@
     blaze-builder            >= 0.3    && < 0.4,
     blaze-builder-enumerator >= 0.2    && < 0.3,
     bytestring               >= 0.9    && < 0.11,
-    case-insensitive         >= 0.3    && < 1.1,
+    case-insensitive         >= 0.3    && < 1.2,
     containers               >= 0.3    && < 0.6,
     enumerator               >= 0.4.13 && < 0.5,
     mtl                      >= 2.0    && < 2.2,
